Remove layout_stride template parameter.

crtrott opened the pull request.

crtrott committed:
Remove layout_stride template parameter.

crtrott commented on compilation_tests/ctest_constexpr_submdspan.cpp:
```diff
@@ -71,7 +72,6 @@ dynamic_extent_1d() {
 
 MDSPAN_STATIC_TEST(dynamic_extent_1d<stdex::layout_left>());
 MDSPAN_STATIC_TEST(dynamic_extent_1d<stdex::layout_right>());
```
these and the following tests won't work, since the constexpr_assert_equal can not be done as a constexpr expression for layout_stride anymore due to the dynamic storage construction. Its probably fixable but I couldn't figure it out ...

crtrott commented on compilation_tests/ctest_no_unique_address.cpp:
```diff
@@ -89,7 +89,7 @@ MDSPAN_STATIC_TEST(
 MDSPAN_STATIC_TEST(
   sizeof(stdex::layout_left::template mapping<
     stdex::extents<42, stdex::dynamic_extent, 73>
```
technically this should have been changed before. Its only those two here I switched from ptrdiff_t to size_t though

crtrott commented on compilation_tests/ctest_no_unique_address.cpp:
```diff
@@ -99,19 +99,11 @@ MDSPAN_STATIC_TEST(
+  >) == 4 * sizeof(size_t)
 );
 
-// Fails on MSVC which adds some padding 
```
this test is simply not true anymore since strides are now always stored dynamically as proposed.

crtrott commented on compilation_tests/ctest_standard_layout.cpp:
```diff
@@ -167,48 +167,35 @@ MDSPAN_STATIC_TEST(
 
 MDSPAN_STATIC_TEST(
   std::is_standard_layout<
-    stdex::layout_stride<12, 1, 3>::template mapping<
```
duplicated tests with other ones now that the template parameter for layout_stride is gone (i.e. only the test will all args dyn_ext need to remain).

crtrott commented on compilation_tests/ctest_standard_layout.cpp:
```diff
@@ -167,48 +167,35 @@ MDSPAN_STATIC_TEST(
   >::value
 );
 
+// TODO: Remove this test alltogether?
```
I think this could still be standard_layout? But maybe not. In either case it doesn't pass on GCC anymore.

crtrott commented on compilation_tests/ctest_trivially_copyable.cpp:
```diff
@@ -167,46 +167,30 @@ MDSPAN_STATIC_TEST(
 
 MDSPAN_STATIC_TEST(
   std::is_trivially_copyable<
-    stdex::layout_stride<12, 1, 3>::template mapping<
```
now duplicated tests are removed, just maintain the ones with were all dynamic extent parameters.

crtrott commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -287,5 +295,19 @@ class extents
 
 };
 
+// CT: Initial implementation for dextents proposed in: P2299
```
I implemented this since I use this as storage for strides. But I put it into detail namespace for now.

dalg24 replied:
```suggestion
// CT: Initial implementation for dextents proposed in: P2299R3
```

crtrott commented on include/experimental/__p0009_bits/layout_stride.hpp:
```diff
@@ -195,38 +187,11 @@ class layout_stride_impl<
-    : __base_t(__base_t{__member_pair_t(e, __strides_storage_t(__construct_psa_from_dynamic_values_tag_t<>{}, strides))})
-  { }
-
-  // TODO @proposal-bug @proposal-extension layout stride needs this constructor
```
I don't need it anymore ...

crtrott commented on tests/test_layout_stride.cpp:
```diff
@@ -74,21 +74,7 @@ using extents_ddd_t = stdex::extents<dyn, dyn, dyn>;
 using zero_stride_maps =
   ::testing::Types<
     layout_stride_case_t<extents_345_t, _ints<>, _ints<dyn, dyn, dyn>, _ints<0, 0, 0>>,
```
removing now duplicated tests.

crtrott commented:
the single header thingy doesn't seem to work though ...

dalg24 commented on compilation_tests/ctest_constexpr_submdspan.cpp:
```diff
@@ -56,9 +56,10 @@ namespace stdex = ::std::experimental;
   int data[] = {1, 2, 3, 4, 5};
   auto s = stdex::basic_mdspan<
-    int, stdex::extents<stdex::dynamic_extent>, Layout>(data, 5);
+    int, ext_t, Layout>(data, 5);
```
Undo

dsunder commented:
Just read over the README and the caveats section needs to be updated.  I believe that this pull request we can remove all the caveats that are listed there.

crtrott committed:
Fix and issue with ussing auto in a wrong place

THis preveneted clang from working

dalg24 commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -208,6 +208,14 @@ class extents
         detail::__construct_psa_from_dynamic_values_tag_t<>{}, dyn}})
   { }
 
+  // Need this constructor for some submdspan implementation stuff
+  // for the layout_stride case where I use an extents object for strides
+  MDSPAN_INLINE_FUNCTION
+  constexpr explicit
+  extents(__storage_t const& sto ) noexcept
+   : __base_t(__base_t{sto})
+  { }
```
I don't like this but I don't have a better idea

crtrott committed:
Addessing review comments

revert one change, move dextents into std::experimental from detail
since we had LEWG review
